### PR TITLE
[ROCm] float16.h should use __HIP__ not USE_ROCM

### PR DESCRIPTION
**Pull request by jeffdaily**
The float16.h header is shared between the CPU and ROCm EPs. The USE_ROCM macro is defined universally, but for the float16.h header we only wish to detect the hip-clang compiler. Otherwise, the CPU EP fails to build because of -Werror -Wuninitialized caused by the USE_ROCM code additions, and the CPU EP should be using a different code path.

### CODE REVIEW

File: include/onnxruntime/core/framework/float16.h [hunk 1]
```diff
@@ -41,11 +41,10 @@ inline bool operator<(const MLFloat16& left, const MLFloat16& right) { return le
 // BFloat16
 struct BFloat16 {
   uint16_t val{0};
-#if defined(USE_ROCM)
-  ORT_HOST_DEVICE BFloat16() = default;
-#else
-  BFloat16() = default;
+#if defined(__clang__) && defined(__HIP__)
+  ORT_HOST_DEVICE
```
Comment (abudup) [inline include/onnxruntime/core/framework/float16.h:45]: Style, nit: Although this avoids duplication, I think the older style improves readability, i.e.,

```
#if defined (__clang__) && defined(__HIP__)
  ORT_HOST_DEVICE BFloat16() = default;
#else
  BFloat16() = default;
#endif
```

File: include/onnxruntime/core/framework/float16.h [hunk 2]
```diff
@@ -41,11 +41,10 @@ inline bool operator<(const MLFloat16& left, const MLFloat16& right) { return le
 // BFloat16
 struct BFloat16 {
   uint16_t val{0};
-#if defined(USE_ROCM)
-  ORT_HOST_DEVICE BFloat16() = default;
-#else
-  BFloat16() = default;
+#if defined(__clang__) && defined(__HIP__)
```
Comment (abudup) [inline include/onnxruntime/core/framework/float16.h:44]: Is it not sufficient to just check for `__HIP__`?

Comment (jeffdaily) [reply include/onnxruntime/core/framework/float16.h:44]: Looks like you're right.  `__HIP__` should be sufficient, based on https://github.com/ROCm-Developer-Tools/HIP/blob/develop/docs/markdown/hip_porting_guide.md#compiler-defines-summary.

**Review by abudup — APPROVED** [review]
Thank you for the fix!

**Pull request merged**